src/api-service/__app__/pool_resize/__init__.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import logging
+import math
+from typing import List
+
+import azure.functions as func
+from onefuzztypes.enums import NodeState, PoolState, ScalesetState
+from onefuzztypes.models import Error
+
+from ..onefuzzlib.pools import Node, Pool, Scaleset
+from ..onefuzzlib.tasks.main import Task
+
+
+def scale_up(pool: Pool, scalesets: List[Scaleset], nodes_needed: int) -> None:
+    logging.info(f"Nodes needed: {nodes_needed}")
+
+    for scaleset in scalesets:
+        if scaleset.state == ScalesetState.running:
+
+            max_size = min(scaleset.max_size(), pool.max_size)
+            logging.info(f"Scaleset size: {scaleset.size}, max_size: {max_size}")
+            if scaleset.size < max_size:
+                current_size = scaleset.size
+                if nodes_needed <= max_size - current_size:
+                    scaleset.size = current_size + nodes_needed
+                    nodes_needed = 0
+                else:
+                    scaleset.size = max_size
+                    nodes_needed = nodes_needed - (max_size - current_size)
+                scaleset.state = ScalesetState.resize
+                scaleset.save()
+
+            else:
+                continue
+
+            if nodes_needed == 0:
+                return
+
+    for _ in range(
+        math.ceil(
+            nodes_needed / min(Scaleset.scaleset_max_size(pool.image), pool.max_size)
+        )
+    ):
+        logging.info(f"Creating Scaleset for Pool {pool.name}")
+        max_nodes_scaleset = min(
+            Scaleset.scaleset_max_size(pool.image), pool.max_size, nodes_needed
+        )
+        scaleset = Scaleset.create(
+            pool_name=pool.name,
+            vm_sku=pool.vm_sku,
+            image=pool.image,
+            region=pool.region,
+            size=max_nodes_scaleset,
+            spot_instances=pool.spot_instances,
+            tags={"pool": pool.name},
+        )
+        scaleset.save()
+        # don't return auths during create, only 'get' with include_auth
+        scaleset.auth = None
+        nodes_needed -= max_nodes_scaleset
+
+
+def scale_down(scalesets: List[Scaleset], nodes_to_remove: int) -> None:
+    for scaleset in scalesets:
+        nodes = Node.search_states(
+            scaleset_id=scaleset.scaleset_id, states=[NodeState.free]
+        )
+        if nodes and nodes_to_remove > 0:
+            max_nodes_remove = min(len(nodes), nodes_to_remove)
+            if max_nodes_remove >= scaleset.size and len(nodes) == scaleset.size:
+                scaleset.state = ScalesetState.halt

```

Comment:
Using shutdown here instead of halt will let nodes that pick up work during your resize process to finish their existing work before getting removed.